javascript/ql/test/library-tests/Promises/flow2.js

```diff
@@ -0,0 +1,17 @@
+(async function () {
+	var source = "source";
+
+	Promise.all([source, "clean"]).then((arr) => {
+		sink(arr); // OK - but flagged by taint-tracking.
+		sink(arr[0]); // NOT OK
+		sink(arr[1]); // OK - but flagged by taint-tracking.
+	})
+
+	var [clean, tainted] = await Promise.all(["clean", source]);
+	sink(clean); // OK - but flagged by taint-tracking
+	sink(tainted); // NOT OK
+
+	var [clean2, tainted2] = await Promise.resolve(Promise.all(["clean", source]));
+	sink(clean2); // OK - but flagged by taint-tracking
+	sink(tainted2); // NOT OK
```

Comment:
Could you add a test case where the inputs to `Promise.all` are promises?

I can't quite picture how this would work. It seems to me we end up with the sequence of steps:
- Store step into promise with `promise-value` pseudo-property
- Store step into array object with array-index property
- Load step from the array object to `Promise.all` with `promise-value` pseudo-property
- Load step from `Promise.all` to array access with array-index property

The above sequence is not reducible to plan data flow because the innermost store/load pair have different property names.

The way I see it, `Promise.all` converts an "array of promises" to a "promise of an array", which in general requires us to swap the order of two properties. This requires two loads steps followed by two store steps (which we unfortunately can't do).

Reply:
> The way I see it, Promise.all converts an "array of promises" to a "promise of an array", which in general requires us to swap the order of two properties. This requires two loads steps followed by two store steps (which we unfortunately can't do).

Your analysis is correct, and we end up not tracking data-flow when the inputs to `Promise.all` are promises.

Reply:
Yeah that's really unfortunate. This PR is a bit of a hard sell with that restriction. Do you mind withholding this PR for a bit while trying out the local heuristic?

Reply:
> Yeah that's really unfortunate. This PR is a bit of a hard sell with that restriction. Do you mind withholding this PR for a bit while trying out the local heuristic?

I think I got it working nicely.  
The last test-case is still not flagged by a `DataFlow::Configuration`, but the `TaintTracking::Configuration` no longer has any FPs in the test.